### PR TITLE
HPCC-17018 Do not lower case string constants in filename conditions

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately

### CODE REVIEW

**Comment by ghalliday** [comment]
Tested by writing a test case to reproduce the problem, and check that the change fixes the issue on the original query.
Also added tracing to look at all the operations that are no longer traversed and check the values for all existing queries to see if anything has been missed (TRIM).
There are a few queries which were previously wrong - tkirk24 and jake37 are notable.
@shamser please review

File: ecl/hql/hqlfold.cpp [hunk 1]
```diff
@@ -7075,15 +7075,41 @@ class LowerCaseTransformer : public NewHqlTransformer
...
+        return completeTransform(expr, args);
     }
 };
 
+/*
+ * The purpose of this function is to normalize a filename as much as possible, so that differences in case for
+ * string constants do not prevent a write to a file, and a subsequent read from the same file to be treated as
+ * the same filenames.  It doesn't aim to be perfect, but to avoid the most common problems.
+ */
 IHqlExpression * lowerCaseHqlExpr(IHqlExpression * expr)
```
Comment (shamser) [inline ecl/hql/hqlfold.cpp]: @ghalliday If this is primarily to normalize, should this function be renamed to reflect purpose - e.g. normalizeFilename?

**Review by shamser — COMMENTED** [review]
@ghalliday The PR title "Do not lower case string constants in filename conditions" may not be reflecting the change implemented.  Should the description be something like "Do not lowercase string constant within comparisons in if/mapto/case"

File: ecl/hql/hqlfold.cpp [hunk 2]
```diff
@@ -7075,15 +7075,41 @@ class LowerCaseTransformer : public NewHqlTransformer
...
+        case no_globalscope:
+        case no_map:
+        case no_trim:
+            return NewHqlTransformer::createTransformed(expr);
+        case no_if:
+        case no_mapto:
+        case no_case:
+            break;
+        default:
+            return LINK(expr);
```
Comment (shamser) [inline ecl/hql/hqlfold.cpp]: @ghalliday should consider changing the name of this the LowerCaseTransformer class to reflect the purpose  - lower-casing is now conditional.

**Comment by ghalliday** [comment]
Need to rerun full regression test before ready to pull

**Comment by ghalliday** [comment]
@shamser please check.  I have run the full regression test.  It also fixed another issue where a #STORED on a computed name wasn't being matched correctly.  (One of the bug*.eclxml examples in the full regression suite, but I have lost the details.)

**Review by shamser — COMMENTED** [review]
@ghalliday The changes fixes bug64452, klgemann42 and skumar5.

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed ok.

**Pull request merged**